File: ReadyTasks/CSharp/ProjectEuler/Task_45/Task_45/Program.cs
using System;

namespace Task_45
{
    class Program
    {
        static bool IsTriangle(long numb)
        {
            // 2 * x == n * n + n
            double D = Math.Sqrt(1 + 8 * numb);
            double x = (-1 + D) / 2;
            return x > 0 && x % 1 == 0;
        }

        static bool IsPentagonal(long numb)
        {
            // 2 * x == 3 * n * n - n
            double D = Math.Sqrt(1 + 24 * numb);
            double x = (1 + D) / 6;
            return x > 0 && x % 1 == 0;
        }

        static bool IsHexagonal(long numb)
        {
            // x == 2 * n * n - n
            double D = Math.Sqrt(1 + 8 * numb);
            double x = (1 + D) / 2;
            return x > 0 && x % 1 == 0;
        }

        static long GetVal(int ind)
        {
            for (long i = 1; ; i++)
            {
                long hexNumb = i * (2 * i - 1);
                if (IsPentagonal(hexNumb) && IsTriangle(hexNumb))
                {
                    if (ind == 0)
                    {
                        return hexNumb;
                    }
                    else
                    {
                        ind--;
                    }
                }
            }
            return -1;
        }

        static void Main(string[] args)
        {
            Console.WriteLine(GetVal(3));
        }
    }
}

File: ReadyTasks/FSharp/ProjectEuler/Task_45/Task_45/Program.fs
open System

let isTriangle numb =
    //x * (x + 1) = 2 * numb => x ^ 2 + x - 2 * numb
    let D = (1L + 8L * numb) |> double |> sqrt
    let res = (-1.0 + D) / 2.0
    res % 1.0 = 0.0

let isPentagonal numb =
    //x * (3 * x - 1) = 2 * numb => 3 * x ^ 2 - x - 2 * numb
    let D = (1L + 24L * numb) |> double |> sqrt
    let res = (1.0 + D) / 6.0
    res % 1.0 = 0.0

let isHexagonal numb =
    //x * (2 * x - 1) = numb => 2 * x ^ 2 - x - numb
    let D = (1L + 8L * numb) |> double |> sqrt
    let res = (1.0 + D) / 4.0
    res % 1.0 = 0.0

let multSeq () =
    Seq.initInfinite (fun x -> ((x + 1) * (2 * x + 1)) |> int64) |> Seq.filter (fun x -> isPentagonal x && isTriangle x)

[<EntryPoint>]
let main argv =
    printfn "%A" (multSeq() |> Seq.take 3)
    0 // return an integer exit code